src/LuaTransformer.ts

```diff
@@ -3095,6 +3153,34 @@ export class LuaTransformer {
                                                   // But this should be changed to retun tstl.createNilLiteral()
                                                   // at some point.
         }
+
+        const symbol = this.checker.getSymbolAtLocation(expression);
+        let symbolId: number | undefined;
+        if (symbol) {
+            if (this.options.noHoisting) {
+                // Check for reference-before-declaration
+                const declaration = tsHelper.getFirstDeclaration(symbol, this.currentSourceFile);
+                if (declaration && expression.pos < declaration.pos) {
+                    throw TSTLErrors.ReferencedBeforeDeclaration(expression);
+                }
+
+            } else {
+                // Track symbols seen in scope
+                if (!this.symbolIds.has(symbol)) {
+                    symbolId = this.genSymbolIdCounter++;
+                    const symbolInfo: SymbolInfo = {symbol, firstSeenAtPos: expression.pos};
+                    this.symbolIds.set(symbol, symbolId);
+                    this.symbolInfo.set(symbolId, symbolInfo);
+                } else {
+                    symbolId = this.symbolIds.get(symbol);
+                }
+                this.scopeStack.forEach(s => {
+                    if (!s.referencedSymbols) { s.referencedSymbols = new Set(); }
+                    s.referencedSymbols.add(symbolId);
+                });
+            }
+        }
+
```

Comment:
Probably nice to refactor this block to its own function with a nice descriptive name.